Units/parser-gdscript.r/sample.d/expected.tags

```diff
@@ -0,0 +1,26 @@
+a	input.gd	/^@export_range(start=0, end=100, step=1) var a = 5$/;"	v	annotations:export_range(start=0, end=100, step=1)
+s	input.gd	/^var s = "Hello"$/;"	v	annotations:export
+arr	input.gd	/^@onready var arr = [1, 2, 3]$/;"	v	annotations:onready
+dict	input.gd	/^var dict = {"key": "value", 2: 3}$/;"	v
+typed_var	input.gd	/^var typed_var: int$/;"	v	typeref:typename:int
+inferred_type	input.gd	/^inferred_type\\$/;"	v	annotations:onready,export_multiline
+ANSWER	input.gd	/^const ANSWER = 42$/;"	v	typeref:typename:const
```

Comment:
This one, `ANSWER` is tagged as a `v`, a variable.
However, I think it should be tagged as a `const` or `constant`.
How do you think?

Reply:
I just followed the example of the first file with a constant variable that I saw in a grep, I went by `parser-cxx.r/cxx11-raw-strings.d/expected.tags`, which has:

```
12:str4	input.cpp	/^static const char* str4 = LR"blah(";int bug4;)blah";$/;"	v	typeref:typename:const char *	file:
13:str5	input.cpp	/^static const char* str5 = u8R"blah(";int bug5;)blah";$/;"	v	typeref:typename:const char *	file:
14:str6	input.cpp	/^static const char* str6 = uR"blah(";int bug6;)blah";$/;"	v	typeref:typename:const char *	file:
15:str7	input.cpp	/^static const char* str7 = UR"blah(";int bug7;)blah";$/;"	v	typeref:typename:const char *	file:
```

Which classifies them with v and puts const in the type, which I followed. GDScript does do a thing where you substitute the keyword `var` entirely with the word `const` to declare a constant, but it's functionally the same as a constant variable I'd think

Reply:
Actually I could drop into godot dev chat real quick and get a thought from there on which to prefer real quick

Reply:
There are developers think differently.

`constant` kind is used in https://github.com/syskrank/vim-gdscript-ctags   .
`C` kind is used for constants in https://github.com/PrestonKnopp/gdtags .

C/C++ parser takes a different approach.  Base on the target language's sematic, the parser recognizes `const` is a part of a type.
```
$ cat /tmp/foo.c
const int i = 1;
$ ./ctags -o - /tmp/foo.c
i	/tmp/foo.c	/^const int i = 1;$/;"	v	typeref:typename:const int
```

ctags is a low-level tool. If possible, ctags should represents different things in different ways. This is one of the mottos of u-ctags.
If you want you can map "const" kind items extracted in ctags to "variable" kind items in geany.

A questionable thing is "const" defined in a method. What kind we should use for it? My idea is just using "const".
Instead, we can delete the "local" kind.  You may think deleting "local" kind conflicts with the motto.
However, it is acceptable because the scope: field of a tag for a local variable can represent the difference.

Let's think about the followig input:
```
$ cat /tmp/foo.gd
var v
const c = 1
func _init():
    var lv = Something.new()
	const lc = 2
```

The current parser output:
```
$ ./ctags --sort=no --kinds-GDScript='*' --fields=+K -o - /tmp/foo.gd
v	/tmp/foo.gd	/^var v$/;"	variable
c	/tmp/foo.gd	/^const c = 1$/;"	variable	typeref:typename:const
_init	/tmp/foo.gd	/^func _init():$/;"	function
lv	/tmp/foo.gd	/^    var lv = Something.new()$/;"	local	function:_init	file:
lc	/tmp/foo.gd	/^	const lc = 2$/;"	local	function:_init	typeref:typename:const	file:
```

My idea:
```
$ ./new-ctags --sort=no --kinds-GDScript='*' --fields=+K -o - /tmp/foo.gd
v	/tmp/foo.gd	/^var v$/;"	variable
c	/tmp/foo.gd	/^const c = 1$/;"	constant
_init	/tmp/foo.gd	/^func _init():$/;"	function
lv	/tmp/foo.gd	/^    var lv = Something.new()$/;"	variable	function:_init	file:
lc	/tmp/foo.gd	/^	const lc = 2$/;"	constant	function:_init	file:
```

Reply:
If the local type gets wiped in favor of metadata I'll have to figure out how to look at metadata to exclude locals from the symbol table in geany, otherwise this is probably fine I guess. Still feel like the relationship between const and var is more akin to C behavior despite the keywording though

Reply:
Here ctags is built from the source code I proposed in #3194.

Consider the following as input:
```
$ cat -n /tmp/foo.gd
     1	var x
     2	func f(x):
     3		var x
     4		func g(x):
     5			var x
     6			
```

ctags-#3194's output:
```
$ u-ctags --sort=no --fields=+n -o - /tmp/foo.gd 
anon_class_49f7b8910100	/tmp/foo.gd	/^var x$/;"	c	line:1
x	/tmp/foo.gd	/^var x$/;"	v	line:1	class:anon_class_49f7b8910100
f	/tmp/foo.gd	/^func f(x):$/;"	m	line:2	class:anon_class_49f7b8910100
x	/tmp/foo.gd	/^func f(x):$/;"	z	line:2	method:anon_class_49f7b8910100.f	file:
x	/tmp/foo.gd	/^	var x$/;"	l	line:3	method:anon_class_49f7b8910100.f	file:
g	/tmp/foo.gd	/^	func g(x):$/;"	f	line:4	method:anon_class_49f7b8910100.f	file:
x	/tmp/foo.gd	/^	func g(x):$/;"	z	line:4	function:anon_class_49f7b8910100.f.g	file:
x	/tmp/foo.gd	/^		var x$/;"	l	line:5	function:anon_class_49f7b8910100.f.g	file:
```
`--fields=+n` option is for attaching `line:` fields:

If we add `--fields=+Z` option to the command line, `scope:` field markers are added. It helps us understand tags file:
```
$ u-ctags --sort=no --fields=+nZ -o - /tmp/foo.gd 
anon_class_49f7b8910100	/tmp/foo.gd	/^var x$/;"	c	line:1
x	/tmp/foo.gd	/^var x$/;"	v	line:1	scope:class:anon_class_49f7b8910100
f	/tmp/foo.gd	/^func f(x):$/;"	m	line:2	scope:class:anon_class_49f7b8910100
x	/tmp/foo.gd	/^func f(x):$/;"	z	line:2	scope:method:anon_class_49f7b8910100.f	file:
x	/tmp/foo.gd	/^	var x$/;"	l	line:3	scope:method:anon_class_49f7b8910100.f	file:
g	/tmp/foo.gd	/^	func g(x):$/;"	f	line:4	scope:method:anon_class_49f7b8910100.f	file:
x	/tmp/foo.gd	/^	func g(x):$/;"	z	line:4	scope:function:anon_class_49f7b8910100.f.g	file:
x	/tmp/foo.gd	/^		var x$/;"	l	line:5	scope:function:anon_class_49f7b8910100.f.g	file:
```

If we add `--fields=+ZK` option to the command line, the long-name of kinds are used. It helps us understand tags file:
```
$ u-ctags --sort=no --fields=+nZK -o - /tmp/foo.gd 
anon_class_49f7b8910100	/tmp/foo.gd	/^var x$/;"	class	line:1
x	/tmp/foo.gd	/^var x$/;"	variable	line:1	scope:class:anon_class_49f7b8910100
f	/tmp/foo.gd	/^func f(x):$/;"	method	line:2	scope:class:anon_class_49f7b8910100
x	/tmp/foo.gd	/^func f(x):$/;"	parameter	line:2	scope:method:anon_class_49f7b8910100.f	file:
x	/tmp/foo.gd	/^	var x$/;"	local	line:3	scope:method:anon_class_49f7b8910100.f	file:
g	/tmp/foo.gd	/^	func g(x):$/;"	function	line:4	scope:method:anon_class_49f7b8910100.f	file:
x	/tmp/foo.gd	/^	func g(x):$/;"	parameter	line:4	scope:function:anon_class_49f7b8910100.f.g	file:
x	/tmp/foo.gd	/^		var x$/;"	local	line:5	scope:function:anon_class_49f7b8910100.f.g	file:
```

The question is how it will be if we use `variable` kind instead of `local`. Let's replace "local" with "variable" with a text editor:
```
anon_class_49f7b8910100	/tmp/foo.gd	/^var x$/;"	class	line:1
x	/tmp/foo.gd	/^var x$/;"	variable	line:1	scope:class:anon_class_49f7b8910100
f	/tmp/foo.gd	/^func f(x):$/;"	method	line:2	scope:class:anon_class_49f7b8910100
x	/tmp/foo.gd	/^func f(x):$/;"	parameter	line:2	scope:method:anon_class_49f7b8910100.f	file:
x	/tmp/foo.gd	/^	var x$/;"	variable	line:3	scope:method:anon_class_49f7b8910100.f	file:
g	/tmp/foo.gd	/^	func g(x):$/;"	function	line:4	scope:method:anon_class_49f7b8910100.f	file:
x	/tmp/foo.gd	/^	func g(x):$/;"	parameter	line:4	scope:function:anon_class_49f7b8910100.f.g	file:
x	/tmp/foo.gd	/^		var x$/;"	variable	line:5	scope:function:anon_class_49f7b8910100.f.g	file:
```

Both using the kind information (parameter or variable) and the scope fields, you can distinguish all `x`.

So I think deleting the `local` kind is acceptable in your use case.
However, keeping is o.k. My original suggestion was introducing "const" kind.